include/hnswlib/space_pq.h

```diff
@@ -0,0 +1,84 @@
+#pragma once
+#include "hnswlib.h"
+#include <memory>
+#include <stdint.h>
+
+namespace hnswlib {
+
+typedef struct pq_dist_param_s {
+  size_t n_subvectors;
+  size_t n_clusters;
+  float *dist_mat;
+} pq_dist_param_t;
+
+template <typename CODETYPE>
+static float PQLookup(const void *pVect1v, const void *pVect2v,
+                      const void *qty_ptr) {
+  CODETYPE *pVect1 = (CODETYPE *)pVect1v;
+  CODETYPE *pVect2 = (CODETYPE *)pVect2v;
+  pq_dist_param_t *qty = (pq_dist_param_t *)qty_ptr;
+  size_t subspace_mat_size = (qty->n_clusters) * (qty->n_clusters);
+  size_t n_clusters = qty->n_clusters;
+  float res = 0;
+
+  for (size_t i = 0; i < qty->n_subvectors; i++) {
+    res += qty->dist_mat[i * subspace_mat_size + (*pVect1) * n_clusters +
+                         (*pVect2)];
+    pVect1++;
+    pVect2++;
+  }
+  return (res);
+}
+
+template <typename CODETYPE> class PQ_L2Space : public SpaceInterface<float> {
+  DISTFUNC<float> fstdistfunc_;
+  size_t data_size_, d_subvectors;
+  pq_dist_param_t param;
+  std::shared_ptr<float *> codebook;
```

Comment:
this shared_ptr as mentioned in other PRs is not clear. If u can use `std::shared_ptr` means u can use `std::array` and this is what u should have.